hwtracer/src/perf/mod.rs

```diff
@@ -3,9 +3,9 @@ use std::alloc::Layout;
 
 pub(crate) mod collect;
 
-const PERF_DFLT_DATA_BUFSIZE: size_t = 64; // Pages
-const PERF_DFLT_AUX_BUFSIZE: size_t = 64 * 1024 * 1024; // MiB
-const PERF_DFLT_INITIAL_TRACE_BUFSIZE: size_t = 1024 * 1024; // MiB
+const PERF_DFLT_DATA_BUFSIZE: size_t = 4; // Pages
```

Comment:
I'm not certain it was correct to change the first two buffer sizes (which from memory are ring buffers unrelated the the total size of a trace). I will have to dig into the perf docs tomorrow.

Reply:
I was hoping you would tell me if this is a good idea or not!

Reply:
OK, so `DATA_BUFSIZE` and `AUX_BUFSIZE` are used for low-level linux-perf interfacing. These are the buffers that the kernel will fill during trace recording, and we get a notification when they are getting full, at which point we have to copy the data out and update the tail pointer (they are conceptually ring buffers).

The AUX buffer is the one containing the PT packets. The DATA buffer contains side-band information (and we don't use or need that for our purposes).

The size of these buffers isn't related to the overall size of a trace. By making them smaller, you only increase the chance of an overflow. I'd suggest putting them back for now (unless it causes a performance problem).

(Note to self, check if we "consume" the DATA buffer, even if we don't use it, because failing to do so can still cause an overflow).

Reply:
(Reading: https://github.com/intel/libipt/blob/master/doc/howto_capture.md)

Reply:
> By making them smaller, you only increase the chance of an overflow

That's exactly what I intended!

Reply:
Why would you want that though?

Suppose the AUX buffer is 1MB big and we are willing to records traces of up to 64MB.

All it would take is a little system load for an underflow during the first 1MB of data.

Reply:
I also notice that the docs say:

> Note that the size of both the AUX and the DATA area has to be a power of two pages. The DATA area needs one additional page to contain the perf_event_mmap_page.

The AUX buffer comment says `MiB` not `pages` so we are probably allocating a lot more than we thought.

Reply:
A PT trace that's really big is likely to produce an AOT trace far longer than we want to compile.

Reply:
Right, but the length of the PT trace is nothing to do with the AUX buffer.

It's a bit like in your audio player, you can configure the buffer size, but that doesn't dictate the length of the tracks you can play in any way. If you make the buffer too short, you will get skips as the buffer underruns.

Reply:
Let me investigate the bogus `MiB` unit.

Reply:
- Revert the data and aux buffer sizes.
 - Add comments to all of the buffer size consts describing them. I can help with this if you are not sure.

About the MiB unit, I see what happens now. We convert the MiBs into pages, then later there is a runtime check that this is a power of two. So it's actually OK, if unexpected.

Reply:
One thing that is a bit odd though is that whilst the AUX bufsize in MiB is OK for a default page size (4096), it can't be suitable for arbitrary page sizes. I wonder if we should round up to the next power of two pages (separate PR)?

Reply:
Not sure. Do I really want PERF_DFLT_DATA_BUFSIZE to be 8x bigger than PERF_DFLT_INITIAL_TRACE_BUFSIZE?

Reply:
I've just reverted it to what it was. I thought that was the plan?

Neither of the `PERF_*` bits have anything to do with the length of a trace, so we shouldn't be changing them in this PR.

Reply:
Let's take a step back. What I wanted to do by reducing these values was causing "too long" traces to result in a "PT overflow". Where by "too long" I meant "much shorter than what we used to consider too long". I don't know how each of these values interacts though, so perhaps the values you're suggesting do what I want?

Reply:
I understand, but reducing `PERF_DFLT_DATA_BUFSIZE` and `PERF_DFLT_AUX_BUFSIZE` isn't the way to do it. Those buffers could be very small and still rarely result in PT overflows. It depends how quickly we service a request to empty the ring buffers. Making them smaller only makes it less likely to overflow, but gives no guarantees. How likely would also be system dependent...